### PR TITLE
Skip auto-loading PSReadLine on Windows if the NVDA screen reader is active

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Skip auto-loading PSReadLine on Windows if the [NVDA screen reader](https://www.nvaccess.org/download/) is active.
The default rendering and formatting on Windows works with NVDA properly.

## PR Context

There is an accessibility issue with PowerShell Core:
> NVDA does not read the text (except first alphabet) whatever we type in the editor but when we delete it by using backspace, it reads all the characters.

PSReadLine moves the cursor a lot when rendering the text, and when a user is editing a multi-line text on screen, it has to re-render all text below the editing position because the rest of the text may need to move forward or backward depending on the editing. So as of today, the PSReadLine is not very friendly to the accessibility aspect of the user experience.
When it comes to `PSReadLine 3.0`, we will design with the accessibility aspect in mind from the very beginning.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4659
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Review by SteveL-MSFT — APPROVED** [review]

**Review by iSazonov — APPROVED** [review]
Should we add tests?

**Comment by SteveL-MSFT** [comment]
Seems like we could call SystemParametersInfo() with SPI_SETSCREENREADER to true to emulate a screen reader as a test case

**Comment by daxian-dbw** [comment]
Sounds doable. I will try adding a test.

**Comment by daxian-dbw** [comment]
Test added based on `SPI_SETSCREENREADER`.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 1]
```diff
@@ -23,8 +23,8 @@
 using System.Runtime.InteropServices;
```
Comment (PoshChan) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: @daxian-dbw, your last commit had 6 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Validate start of console host.Error occurred in Describe block
```powershell
Exception calling "DeactivateScreenReader" with "0" argument(s): "Unable to load shared library 'user32.dll' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: dlopen(libuser32.dll, 1): image not found"
At /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Host/ScreenReader.Tests.ps1:37 char:9
```
Import-Module should be case insensitive.Import-Module can import a module using different casing using 'TESTMODULE/1.1' and manifest:$true
```powershell
The term 'mytest' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1: line 298
298:         mytest | Should -BeExactly "hello"

```
Basic Alias Provider Tests.Validate basic PSDrive Cmdlets.Create a new PSDrive
```powershell
Expected exactly 'NewDifferentPSDrive', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/PSDrive.Tests.ps1: line 23
23:                 $newDrive.Name | Should -BeExactly "NewDifferentPSDrive"
```
Basic Alias Provider Tests.Validate basic PSDrive Cmdlets.Read data from a PSDrive
```powershell
Expected exactly 'PsTestDriveName', but got $null.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/PSDrive.Tests.ps1: line 33
33:             $driveProp.PSDrive.Name | Should -BeExactly $psDriveName
```
Basic Alias Provider Tests.Validate basic PSDrive Cmdlets.Remove the PSDrive
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/vsts/agent/2.155.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/PSDrive.Tests.ps1: line 40
40:             $existsBefore | Should -BeTrue
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: @daxian-dbw, your last commit had 6 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Validate start of console host.Error occurred in Describe block
```powershell
Exception calling "DeactivateScreenReader" with "0" argument(s): "Unable to load shared library 'user32.dll' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libuser32.dll: cannot open shared object file: No such file or directory"
At /home/vsts/work/1/s/test/powershell/Host/ScreenReader.Tests.ps1:37 char:9
```
Import-Module should be case insensitive.Import-Module can import a module using different casing using 'TESTMODULE/1.1' and manifest:$true
```powershell
The term 'mytest' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Core/Import-Module.Tests.ps1: line 298
298:         mytest | Should -BeExactly "hello"

```
Basic Alias Provider Tests.Validate basic PSDrive Cmdlets.Create a new PSDrive
```powershell
Expected exactly 'NewDifferentPSDrive', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/PSDrive.Tests.ps1: line 23
23:                 $newDrive.Name | Should -BeExactly "NewDifferentPSDrive"
```
Basic Alias Provider Tests.Validate basic PSDrive Cmdlets.Read data from a PSDrive
```powershell
Expected exactly 'PsTestDriveName', but got $null.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/PSDrive.Tests.ps1: line 33
33:             $driveProp.PSDrive.Name | Should -BeExactly $psDriveName
```
Basic Alias Provider Tests.Validate basic PSDrive Cmdlets.Remove the PSDrive
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/PSDrive.Tests.ps1: line 40
40:             $existsBefore | Should -BeTrue
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: @daxian-dbw, your last commit had 1 failures in `PowerShell-CI-linux`
Invoke-WebRequest tests.Denial of service.Image Parsing
```powershell
Expected the actual value to be greater than 10, but got 0.339536922042311.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1: line 1897
1897:             $pathologicalRatio | Should -BeGreaterThan 10
```

**Review by iSazonov — APPROVED** [review]

**Review by SteveL-MSFT — APPROVED** [review]

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs [hunk 2]
```diff
@@ -55,6 +55,11 @@ internal sealed partial class ConsoleHost
         internal const int ExitCodeCtrlBreak = 128 + 21; // SIGBREAK
         internal const int ExitCodeInitFailure = 70; // Internal Software Error
         internal const int ExitCodeBadCommandLineParameter = 64; // Command Line Usage Error
+        private const uint SPI_GETSCREENREADER = 0x0046;
+
+        [DllImport("user32.dll", SetLastError = true)]
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: We could remove `SetLastError = true` - we don't use it.

Comment (daxian-dbw) [reply src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs]: I prefer to keep it, even it's not used in this case. People copy code, and this is a good practice for PInvoke.

**Pull request merged**